### PR TITLE
Try uncollapsing margins on the columns block.

### PR DESCRIPTION
**Pull request by jasmussen**
This is a first stab at fixing #7766.

It does not address it fully, I believe further work is necessary, including that of #7694. However this should at least improve the collapsing of the parent container.

In the following screenshots, I've selected all blocks, as this highlights using transparency and overlay effects, the _boundaries_ and how they overlap. Here's what's in `master`:

before

Here's with this PR applied:

after

As you can see, there are zeroed out margins on the first and last container children inside the columns block. This is there to make the columns block flow more directly with the surrounding content, but this can be revisited, and as noted, #7694 might help here.

Noting that if we were to _remove_ zeroed out margins on the first and last container children, here's what it would look like:

no collapsing margins

### CODE REVIEW

**Review by aduth — APPROVED** [review]
This seems like a small improvement over master, at least as far as overlapping goes. We should stop considering this as a "Columns" issue, and address it as generally affecting any block with inner blocks. In that sense, these changes do nothing for any other block with inner blocks.

File: core-blocks/columns/editor.scss [hunk 1]
```diff
@@ -32,12 +32,17 @@
...
+// This block has flex container children. Flex container margins do not collapse: https://www.w3.org/TR/css-flexbox-1/#flex-containers.
 // Therefore, let's at least not add any additional margins here.
+// The goal is for the editor to look more like the front-end.
 .editor-block-list__layout .editor-block-list__block[data-type="core/columns"] > .editor-block-list__block-edit,
 .editor-block-list__layout .editor-block-list__block[data-type="core/column"] > .editor-block-list__block-edit {
 	margin-top: 0;
 	margin-bottom: 0;
+
+	// This uncollapses margins on this parent container.
+	padding-top: 1px;
```
Comment (aduth) [inline core-blocks/columns/editor.scss]: Would `0.1px` be a better value here? While it's not valid in the sense of a fraction of a pixel, it achieves the same effect while not having a visual difference (since it will round to `0`).

**Comment by jasmussen** [comment]
> We should stop considering this as a "Columns" issue, and address it as generally affecting any block with inner blocks. In that sense, these changes do nothing for any other block with inner blocks.

Step one of making margins collapse was to ensure that the block "foot print" was accurate in the backend, compared to the front-end, specifically to ensure that the UI we paint around each block — borders, multi select squares, toolbars, side UI — did not affect the foot print at all.

Step two is actually adding margins to each and every block, **on a per-block basis**. Essentially, like a front-end styleseheet, a paragraph needs one margin, a list needs another. Perhaps a columns block doesn't need any margins at all. 

Right now we are applying the same margins in a blanket statement, to all blocks. This happens here: https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/block-list/style.scss#L140 

Removing that rule removes any artificial margins added to all the blocks (with the added notes that we have some negative margins to compensate for this). 

In other words, I would suggest that the next step in improving this, and addressing your comment, is to remove those artificial margins, and then rebuild them block by block. Do you agree this is the right approach?
